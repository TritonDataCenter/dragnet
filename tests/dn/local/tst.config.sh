#!/bin/bash

#
# tst.config.sh: tests data source configuration
#

. $(dirname $0)/../common.sh

tmpfile=/var/tmp/$(basename $0).$$
echo "using tmpfile $tmpfile" >&2

md5cmd=""
expmd5=""

function rundn
{
	echo "# dn" "$@"
	DRAGNET_CONFIG=$tmpfile dn "$@"
	status=$?
	echo
	return $status
}

function shouldfail
{
	if "$@" 2>&1 | head -3; then
		echo "didn't expect that to succeed!" >&2
		exit 1
	fi

	return 0
}

if [[ -f $HOME/.dragnetrc ]]; then
	md5cmd="$(which md5)"
	[[ ! -f "$md5cmd" ]] && md5cmd="$(which md5sum)"
	expmd5="$($md5cmd < $HOME/.dragnetrc)"
fi

set -o errexit
set -o pipefail

#
# Data sources
#

# Dump initial state
rundn datasource-list
rundn datasource-list -v

# Fail: no path
shouldfail rundn datasource-add junk3
# Fail: bad filter
shouldfail rundn datasource-add junk3 --filter='{' --path=/junk

# Basic add
rundn datasource-add junk --path=/junk
# Add with filter
rundn datasource-add junk2 --path=/junk \
    --filter='{ "eq": [ "req.method", "GET" ] }'
# Dump state
rundn datasource-list
rundn datasource-list -v
rundn datasource-show junk
rundn datasource-show -v junk

# Fail: duplicate name
shouldfail rundn datasource-add junk --path=/junk

# Update
rundn datasource-update junk2 --backend=manta --path=/foo/bar \
    --index-path=/bar/foo --filter={} --data-format=json-skinner \
    --time-format=%Y --time-field=foo
rundn datasource-show junk2
rundn datasource-show -v junk2
shouldfail rundn datasource-update
shouldfail rundn datasource-update nonexistent

# Clean up and dump state
rundn datasource-remove junk2
rundn datasource-list
rundn datasource-list -v

rundn datasource-remove junk
rundn datasource-list
rundn datasource-list -v

# Fail: datasource not present
shouldfail rundn datasource-remove junk

# Add Manta-based dataset with more parameters
rundn datasource-add manta-based --backend=manta --path=/junk
rundn datasource-add manta-based2 --backend=manta --path=/junk \
    --time-format=%Y/%m/%d/%H --data-format=json-skinner
rundn datasource-list
rundn datasource-list -v

#
# Metrics
#

# Dump initial state
rundn metric-list manta-based
rundn metric-list manta-based2
rundn metric-list -v manta-based
rundn metric-list -v manta-based2

# Fail: bad filter
shouldfail rundn metric-add --filter={ manta-based met1
# Fail: no datasource
shouldfail rundn metric-add met1

# Basic add
rundn metric-add manta-based met1
rundn metric-list manta-based
rundn metric-list -v manta-based

# Add with filter and breakdowns
rundn metric-add --filter='{ "eq": [ "req.method", "GET" ] }' manta-based met2
rundn metric-add --filter='{ "eq": [ "req.method", "GET" ] }' \
    --breakdowns=host,req.method,latency[aggr=quantize] manta-based met3
rundn metric-list manta-based
rundn metric-list -v manta-based

# Fail: duplicate name
shouldfail rundn metric-add manta-based met1

rundn metric-remove manta-based met1
rundn metric-remove manta-based met2
rundn metric-remove manta-based met3
shouldfail rundn metric-remove manta-based met2

# Clean up and dump state.
rundn datasource-remove manta-based2
rundn datasource-remove manta-based
rundn datasource-list
rundn datasource-list -v

rm -f $tmpfile

if [[ -n "$expmd5" ]]; then
	if [[ "$($md5cmd < $HOME/.dragnetrc)" != "$expmd5" ]]; then
		echo "oops! clobbered .dragnetrc!" >&2
		exit 1
	fi
fi
