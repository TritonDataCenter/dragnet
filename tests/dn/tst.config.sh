#!/bin/bash

#
# tst.config.sh: tests data source configuration
#

tmpfile=/var/tmp/$(basename $0).$$
echo "using tmpfile $tmpfile" >&2

md5cmd=""
expmd5=""

function rundn
{
	echo "# dn" "$@"
	DRAGNET_CONFIG=$tmpfile dn "$@"
	echo
}

if [[ -f $HOME/.dragnetrc ]]; then
	md5cmd="$(which md5)"
	[[ ! -f "$md5cmd" ]] && md5cmd="$(which md5sum)"
	expmd5="$($md5cmd < $HOME/.dragnetrc)"
fi

set -o errexit
rundn datasource-list
rundn datasource-list -v

rundn datasource-add junk --path=/junk
rundn datasource-list
rundn datasource-list -v

rundn datasource-remove junk
rundn datasource-list
rundn datasource-list -v

rundn datasource-add manta-based --backend=manta --path=/junk
rundn datasource-list
rundn datasource-list -v

rundn datasource-add manta-based2 --backend=manta --path=/junk \
    --time-format=%Y/%m/%d/%H --data-format=json-skinner
rundn datasource-list
rundn datasource-list -v

rundn datasource-remove manta-based2
rundn datasource-list
rundn datasource-list -v

rundn datasource-remove manta-based
rundn datasource-list
rundn datasource-list -v

rm -f $tmpfile

if [[ -n "$expmd5" ]]; then
	if [[ "$($md5cmd < $HOME/.dragnetrc)" != "$expmd5" ]]; then
		echo "oops! clobbered .dragnetrc!" >&2
		exit 1
	fi
fi
