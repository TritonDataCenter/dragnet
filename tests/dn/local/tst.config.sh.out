# dn datasource-list
DATASOURCE           LOCATION                                                   

# dn datasource-list -v
DATASOURCE           LOCATION                                                   

# dn datasource-add junk3
dn: "path" option is required
usage: dn SUBCOMMAND [OPTIONS] ARGS
# dn datasource-add junk3 --filter={ --path=/junk
dn: invalid filter: Unexpected end of input
usage: dn SUBCOMMAND [OPTIONS] ARGS
# dn datasource-add junk --path=/junk

# dn datasource-add junk2 --path=/junk --filter={ "eq": [ "req.method", "GET" ] }

# dn datasource-list
DATASOURCE           LOCATION                                                   
junk                 file://junk                                                
junk2                file://junk                                                

# dn datasource-list -v
DATASOURCE           LOCATION                                                   
junk                 file://junk                                                
    dataFormat: "json"
junk2                file://junk                                                
    filter:     {"eq":["req.method","GET"]}
    dataFormat: "json"

# dn datasource-show junk
DATASOURCE           LOCATION                                                   
junk                 file://junk                                                

# dn datasource-show -v junk
DATASOURCE           LOCATION                                                   
junk                 file://junk                                                
    dataFormat: "json"

# dn datasource-add junk --path=/junk
dn: datasource "junk" already exists

# dn datasource-update junk2 --backend=manta --path=/foo/bar --index-path=/bar/foo --filter={} --data-format=json-skinner --time-format=%Y --time-field=foo

# dn datasource-show junk2
DATASOURCE           LOCATION                                                   
junk2                manta://us-east.manta.joyent.com/foo/bar                   

# dn datasource-show -v junk2
DATASOURCE           LOCATION                                                   
junk2                manta://us-east.manta.joyent.com/foo/bar                   
    filter:     {}
    dataFormat: "json-skinner"
    indexPath:  "/bar/foo"
    timeFormat: "%Y"
    timeField:  "foo"

# dn datasource-update
dn: missing arguments
usage: dn SUBCOMMAND [OPTIONS] ARGS
# dn datasource-update nonexistent
dn: datasource "nonexistent" does not exist

# dn datasource-remove junk2

# dn datasource-list
DATASOURCE           LOCATION                                                   
junk                 file://junk                                                

# dn datasource-list -v
DATASOURCE           LOCATION                                                   
junk                 file://junk                                                
    dataFormat: "json"

# dn datasource-remove junk

# dn datasource-list
DATASOURCE           LOCATION                                                   

# dn datasource-list -v
DATASOURCE           LOCATION                                                   

# dn datasource-remove junk
dn: datasource "junk" does not exist

# dn datasource-add manta-based --backend=manta --path=/junk

# dn datasource-add manta-based2 --backend=manta --path=/junk --time-format=%Y/%m/%d/%H --data-format=json-skinner

# dn datasource-list
DATASOURCE           LOCATION                                                   
manta-based          manta://us-east.manta.joyent.com/junk                      
manta-based2         manta://us-east.manta.joyent.com/junk                      

# dn datasource-list -v
DATASOURCE           LOCATION                                                   
manta-based          manta://us-east.manta.joyent.com/junk                      
    dataFormat: "json"
manta-based2         manta://us-east.manta.joyent.com/junk                      
    dataFormat: "json-skinner"
    timeFormat: "%Y/%m/%d/%H"

# dn metric-list manta-based
DATASOURCE           METRIC              

# dn metric-list manta-based2
DATASOURCE           METRIC              

# dn metric-list -v manta-based
DATASOURCE           METRIC              

# dn metric-list -v manta-based2
DATASOURCE           METRIC              

# dn metric-add --filter={ manta-based met1
dn: invalid filter: Unexpected end of input
usage: dn SUBCOMMAND [OPTIONS] ARGS
# dn metric-add met1
dn: missing arguments
usage: dn SUBCOMMAND [OPTIONS] ARGS
# dn metric-add manta-based met1

# dn metric-list manta-based
DATASOURCE           METRIC              
manta-based          met1                

# dn metric-list -v manta-based
DATASOURCE           METRIC              
manta-based          met1                

# dn metric-add --filter={ "eq": [ "req.method", "GET" ] } manta-based met2

# dn metric-add --filter={ "eq": [ "req.method", "GET" ] } --breakdowns=host,req.method,latency[aggr=quantize] manta-based met3

# dn metric-list manta-based
DATASOURCE           METRIC              
manta-based          met1                
manta-based          met2                
manta-based          met3                

# dn metric-list -v manta-based
DATASOURCE           METRIC              
manta-based          met1                
manta-based          met2                
    filter:     {"eq":["req.method","GET"]}
manta-based          met3                
    filter:     {"eq":["req.method","GET"]}
    breakdowns: host, req.method, latency

# dn metric-add manta-based met1
dn: metric "met1" already exists

# dn metric-remove manta-based met1

# dn metric-remove manta-based met2

# dn metric-remove manta-based met3

# dn metric-remove manta-based met2
dn: datasource "manta-based" metric "met2" does not exist

# dn datasource-remove manta-based2

# dn datasource-remove manta-based

# dn datasource-list
DATASOURCE           LOCATION                                                   

# dn datasource-list -v
DATASOURCE           LOCATION                                                   

